pkg/operator/genericoperatorclient/dynamic_operator_client.go

```diff
@@ -156,6 +190,76 @@ func (c dynamicOperatorClient) UpdateOperatorStatus(resourceVersion string, stat
 	return retStatus, nil
 }
 
+func (c dynamicOperatorClient) AddFinalizer(finalizer string) error {
+	uncastInstance, err := c.informer.Lister().Get(c.configName)
+	if err != nil {
+		if errors.IsNotFound(err) && c.fakeMissingInstance {
+			return nil
+		}
+		return err
+
+	}
+
+	instance := uncastInstance.(*unstructured.Unstructured)
+	finalizers := instance.GetFinalizers()
+	for _, f := range finalizers {
+		if f == finalizer {
+			return nil
+		}
+	}
+
+	// Change is needed
+	klog.V(4).Infof("Adding finalizer %q", finalizer)
+	newFinalizers := append(finalizers, finalizer)
+	err = c.saveFinalizers(instance, newFinalizers)
+	if err != nil {
+		return err
+	}
+	klog.V(2).Infof("Added finalizer %q", finalizer)
+	return err
+}
+
+func (c dynamicOperatorClient) RemoveFinalizer(finalizer string) error {
+	uncastInstance, err := c.informer.Lister().Get(c.configName)
+	if err != nil {
+		if errors.IsNotFound(err) && c.fakeMissingInstance {
+			return nil
+		}
+		return err
+
+	}
+
+	instance := uncastInstance.(*unstructured.Unstructured)
+	finalizers := instance.GetFinalizers()
+	found := false
+	newFinalizers := make([]string, 0, len(finalizers))
+	for _, f := range finalizers {
+		if f == finalizer {
+			found = true
+			continue
+		}
+		newFinalizers = append(newFinalizers, f)
+	}
+	if !found {
+		return nil
+	}
+
+	klog.V(42).Infof("Removing finalizer %q: %v", finalizer, newFinalizers)
+	err = c.saveFinalizers(instance, newFinalizers)
+	if err != nil {
+		return err
+	}
+	klog.V(2).Infof("Removed finalizer %q", finalizer)
+	return nil
+}
+
+func (c dynamicOperatorClient) saveFinalizers(instance *unstructured.Unstructured, finalizers []string) error {
```

Comment:
this looks wrong. You'll be rewriting finalizers if someone adds in between.  We made an update helper to that can take a function to handle the mutation.

Reply:
Should it get Conflict error then and the operator re-tries shortly?

I did not notice any special logic when removing finalziers elsewhere in library-go
https://github.com/openshift/library-go/blob/17708d019f9cb170ff1f31265d5acfa82b9a82e9/pkg/operator/encryption/controllers/prune_controller.go#L154-L158

Or do you mean `RetryOnConflict` loop?
https://github.com/openshift/library-go/blob/17708d019f9cb170ff1f31265d5acfa82b9a82e9/pkg/operator/secretspruner/prune_controller.go#L86-L87